### PR TITLE
File-like objects support

### LINKED ISSUE #35: Support filelike objects, not only pathes
Hi!
I'm trying to use this library with remote files defined as file-likes in python.
`mutagen` supports file-like interface, why tho mediafile doesn't?
Docs say it needs local file path as first argument.

### PR DESCRIPTION
**Pull request by JuniorJPDJ**
and simple and useful `MediaFile.as_dict()` function

Resolves #35

### CODE REVIEW

**Comment by sampsyo** [comment]
Thanks for getting this started! If you're interested in changing `path` to `filename`, perhaps we should do that in a separate PR so it's easier to review the core of the "file-like object" feature separately? (In general, for me, I find it's easier to fully understand one change at a time.)

**Comment by JuniorJPDJ** [comment]
It's connected as in case of FileThing it doesn't have to be `path` already but it will always be some sort of file name.
I can split it to separate commits for you if it's gonna make review easier ;D

**Comment by sampsyo** [comment]
That would be awesome, if you don't mind!

**Comment by JuniorJPDJ** [comment]
Done!
If possible - I'd like to get your review and merge on #42 first tho.

**Comment by JuniorJPDJ** [comment]
Rebased on top of master.

**Review by sampsyo — COMMENTED** [review]
A few miscellaneous comments within!

File: test/test_mediafile_edge.py [hunk 1]
```diff
@@ -436,7 +436,6 @@ def test_write(self):
         shutil.copy(src, path)
         mf = mediafile.MediaFile(path)
         mf.read_only_field = "something terrible"
-        mf.path = os.path.join(self.temp_dir, b'test.flac')
```
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply test/test_mediafile_edge.py:439]:
```python
    def test_write(self):
        src = os.path.join(_common.RSRC, b'empty.flac')
        path = os.path.join(self.temp_dir, b'test.flac')
        shutil.copy(src, path)
        mf = mediafile.MediaFile(path)
        mf.read_only_field = "something terrible"
        mf.path = os.path.join(self.temp_dir, b'test.flac')
        mf.save()
        self.assertNotIn(self.key, mf.mgfile.tags)
```

This is a whole function.
`path` is already set to exactly this path and then handled to `mediafile` in line:

```python
 mf = mediafile.MediaFile(path)
```

So it was setting `mf.path` to `mf.path` ;D

File: mediafile.py [hunk 2]
```diff
@@ -131,15 +135,38 @@ def mutagen_call(action, path, func, *args, **kwargs):
...
+
+        def wrapper(*args, **kwargs):
+            return mutagen_call('loadfile', '', f, *args, **kwargs)
+        return wrapper
+    return wrap
+
 # Utility.
 
+
+def _update_filething(filething):
```
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:160]: The problem is I'm not exactly sure why and where exactly closing happens.
Debugging this issue was guessing game.
`mutagen` was closing my files (only in case of local files, wtf) after reading and I couldn't find any `.close()` call in whole mutagen code nor using debugger on own proxy file-like.
Anyway - this allows to reopen files in `r+` when saving when opened as `r` on loading.
Proxy file-like weren't closed at all..

I'm not sure what could I write in this docstring what's not written in comment anyway.

File: mediafile.py [hunk 3]
```diff
@@ -1518,6 +1549,19 @@ def __init__(self, path, id3v23=False):
         # Set the ID3v2.3 flag only for MP3s.
         self.id3v23 = id3v23 and self.type == 'mp3'
 
+    @property
+    def filesize(self):
```
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:1588]: It's pretty obvious for me, but if you say it's needed.. ;D

File: mediafile.py [hunk 4]
```diff
@@ -1468,20 +1495,24 @@ class MediaFile(object):
     """Represents a multimedia file on disk and provides access to its
     metadata.
     """
-    def __init__(self, path, id3v23=False):
+    @loadfile()
+    def __init__(self, filething, id3v23=False):
```
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:1534]: I'm not sure if I can say what exactly is it without scaring users ;/

`mutagen` itself just [mentions](https://mutagen.readthedocs.io/en/latest/user/filelike.html) that you can put string or filelike as first argument.
Problem is that technically this [first argument is eaten by `loadfile` decorator](https://github.com/quodlibet/mutagen/blob/30f373fa6d56e1afa17d48a0c6f3e111267fbd32/mutagen/_util.py#L141) and it's not `filething` from func definition. `filething` from func definition is [`namedtuple`](https://github.com/quodlibet/mutagen/blob/30f373fa6d56e1afa17d48a0c6f3e111267fbd32/mutagen/_util.py#L195) from `mutagen._utils`.

So TLDR:
I'm not sure how to document this.
Do you have any idea?

File: mediafile.py [hunk 5]
```diff
@@ -131,15 +135,38 @@ def mutagen_call(action, path, func, *args, **kwargs):
     except mutagen.MutagenError as exc:
         log.debug(u'%s failed: %s', action, six.text_type(exc))
         raise UnreadableFileError(path, six.text_type(exc))
+    except UnreadableFileError:
+        # reraise our errors without changes
+        raise
```
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:140]: `__init__` function is now whole wrapped by `loadfile` decorator, which calls `mutagen._util.loadfile` through `mutagen_call` (original loadfile also raises exceptions) so we.. use whole `__init__` function through `mutagen_call` and then call it inside, so exceptions are caught by it two times.

File: mediafile.py [hunk 6]
```diff
@@ -131,15 +135,38 @@ def mutagen_call(action, path, func, *args, **kwargs):
...
     except Exception as exc:
         # Isolate bugs in Mutagen.
         log.debug(u'%s', traceback.format_exc())
         log.error(u'uncaught Mutagen exception in %s: %s', action, exc)
         raise MutagenError(path, exc)
 
 
+def loadfile(method=True, writable=False, create=False):
```
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:148]: Ok ;D
Should be pretty easy.

File: mediafile.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:140]: I'm now, after some time, considering removing `mutagen_call` from there and catching this exceptions myself in decorator.
It probably would be cleaner solution...

File: test/test_mediafile_edge.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sampsyo) [reply test/test_mediafile_edge.py:439]: Huh; interesting! I wonder why this was here in the first place.

File: mediafile.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sampsyo) [reply mediafile.py:1534]: I think I'd just say that `filething` can either be a string or a file-like object.

File: mediafile.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:140]: It's not cleaner solution as it finished almost as copy-pasta of `mutagen_call`.
I'll just add one more comment here and leave it as it was in previous version.

File: mediafile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply mediafile.py:160]: I found the reason!
`mutagen._util._openfile` [closes files it opened](https://github.com/quodlibet/mutagen/blob/30f373fa6d56e1afa17d48a0c6f3e111267fbd32/mutagen/_util.py#L273) just after [function needing file finishes](https://github.com/quodlibet/mutagen/blob/30f373fa6d56e1afa17d48a0c6f3e111267fbd32/mutagen/_util.py#L155).

**Comment by JuniorJPDJ** [comment]
I think I fixed most of issues you asked about ;D

**Comment by JuniorJPDJ** [comment]
I also re-added `MediaFile.path` as `@property` - I don't want to completely break backwards compatibility.

**Comment by JuniorJPDJ** [comment]
.

**Comment by sampsyo** [comment]
Just returning to this PR after a long delay—is everything in order from your perspective, @JuniorJPDJ? Maybe we should add something to the documentation about this new capability so people know they can use it with open files?

File: mediafile.py [hunk 7]
```diff
@@ -1503,20 +1540,24 @@ class MediaFile(object):
...
     metadata.
     """
-    def __init__(self, path, id3v23=False):
-        """Constructs a new `MediaFile` reflecting the file at path. May
-        throw `UnreadableFileError`.
+    @loadfile()
+    def __init__(self, filething, id3v23=False):
+        """Constructs a new `MediaFile` reflecting the provided file.
+        May throw `UnreadableFileError`.
+        First argument can be a string (file path) or filelike object.
```
Comment (JuniorJPDJ) [inline mediafile.py:1547]: There's something about it.
IMO it's enough.

**Comment by JuniorJPDJ** [comment]
IMO it's enough. @up is line describing it
If you think it's not enough - tell me please what exactly would you like to see.

**Comment by JuniorJPDJ** [comment]
bump

**Comment by sampsyo** [comment]
Thanks for your patience! This all looks great. I have updated some documentation and will merge now. :sparkles:
